### PR TITLE
[4.0] Rewrite indexer in vanilla Javascript

### PR DESCRIPTION
**Pull request by C-Lodder**
### Summary of Changes

Remove all jQuery from `indexer.js` (apart from 1 line which I'm not sure about)

### Testing Instructions

Ensure the smart search indexer still works.

There is an open issue (https://github.com/joomla/joomla-cms/issues/18861) by which the index doesn't work with sample data on a multilang site. So testing can be done on a installation with no sample data.

@franz-wohlkoenig Can you check if this solves the issue you raised? Don't think it will but just incase.

### CODE REVIEW

File: media/com_finder/js/indexer.js [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
-	var batchSize= null;
-	var offset= null;
-	var progress= null;
-	var optimized= false;
-	var path = 'index.php?option=com_finder&tmpl=component&format=json';
-
-	var initialize = function () {
-		offset = 0;
+var FinderIndexer = function() {
+
```
Comment (dgrammatiko) [inline media/com_finder/js/indexer.js]: Can you make this Joomla.finderIndexer() ?

**Comment by ghost** [comment]
can test in a few Days (reminder #18861)

File: media/com_finder/js/indexer.js [hunk 2]
```diff
@@ -1,30 +1,47 @@
...
+			url:       path,
+			method:    'GET',
+			data:      { task: task },
+			perform:   true,
+			headers:   {'Content-Type': 'application/x-www-form-urlencoded'},
+			onSuccess: function(response, xhr) {
+				handleResponse();
+			},
+			onError: function(xhr) {
+				handleFailure();
```
Comment (Fedik) [inline media/com_finder/js/indexer.js]: I think missed `xhr`:
`handleFailure(xhr);`
same for handleResponse

Comment (C-Lodder) [reply media/com_finder/js/indexer.js]: Yeah just noticed. thanks :)

**Comment by C-Lodder** [comment]
Running into `{"error":true,"code":404,"message":"View not found [name, type, prefix]: index, json, "}` at the moment so need to fix.

**Comment by C-Lodder** [comment]
@Fedik @dgt41 I don't suppose you know what the issue is here:

https://github.com/joomla/joomla-cms/pull/18959/files#diff-6ba0e76dee6055600da400d76c8b9badR28

When I used `Joomla.request` (the commented out code) then I get the following error:

> {"error":true,"code":404,"message":"View not found [name, type, prefix]: index, json, "}

**Comment by C-Lodder** [comment]
No worries, fixed

**Comment by C-Lodder** [comment]
Errors fixed

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 7829e5b428b8258cb4568680f4b10e5a3973b6df<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18959">issues.joomla.org/tracker/joomla-cms/18959</a>.</sub>

**Comment by infograf768** [comment]
@C-Lodder 
This does not work here. The modal remains open and indexing never completes.
I get the same error in min and full js

"SyntaxError: JSON.parse: unexpected non-whitespace character after JSON data at line 1 column 2732 of the JSON data"

screen shot 2017-12-11 at 09 05 19

**Comment by C-Lodder** [comment]
Hmm ok same error as before. I see you have debug enabled so I'll test tonight

**Comment by infograf768** [comment]
Also when indexing or simply displaying the Manager I get:

[13-Dec-2017 12:01:10 Europe/Berlin] PHP Notice:  Undefined variable: name in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_finder/Field/ContentmapField.php on line 99
[13-Dec-2017 12:01:10 Europe/Berlin] PHP Notice:  Undefined variable: name in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_finder/Field/ContentmapField.php on line 110
[13-Dec-2017 12:01:10 Europe/Berlin] PHP Notice:  Undefined variable: name in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_finder/Field/ContentmapField.php on line 115

**Comment by C-Lodder** [comment]
@infograf768 Please open a new issue.

I can see why this is happening but I have no idea what the default value for `$name` should be in that file as I've never used com_fields before

**Comment by infograf768** [comment]
com_finder you mean I guess . 😺

**Comment by C-Lodder** [comment]
Oops, yes.

I compared to J3.x and the files looks like it would have the same problem there too. `$name` isn't defined before the `if/else` statement

**Comment by infograf768** [comment]
indeed,  although I don't remember getting the error in 3.x

**Pull request merged**